docs/specs/tsd-003-output-buffer-strategy.md

```diff
@@ -0,0 +1,72 @@
+# Telegraf Output Buffer Strategy
+
+## Objective
+
+Introduce a new agent-level config option to choose a disk buffer strategy for
+output plugin metric queues.
+
+## Overview
+
+Currently when a Telegraf output metric queue fills, either due to incoming
+metrics being too fast or various issues with writing to the output, such as
+connection failures or rate limiting, new metrics are dropped and never
+written to the output. This specification defines a set of options to make
+this output queue more durable by persisting pending metrics to disk rather
+than only an in-memory limited size queue.
+
+## Keywords
+
+output plugins, agent configuration, persist to disk
+
+## Agent Configuration
+
+The configuration is at the agent-level, with options for:
+- **Memory**, the current implementation, with no persistance to disk
+- **Write-through**, all metrics are also written to disk using a
+  Write Ahead Log (WAL) file
+- **Disk-overflow**, when the memory buffer fills, metrics are flushed to a
+  WAL file to avoid dropping overflow metrics
+
+As well as an option to specify a directory to store the WAL files on disk,
+with a default value. These configurations are global, and no change means
+memory only mode, retaining current behavior.
+
+## Metric Ordering and Tracking
+
+Tracking metrics will be accepted either on a successful write to the output
+source like currently, or on write to the WAL file in the case of the
+disk-overflow option. Metrics will be written to their appropriate output in
```

Comment:
We would accept tracking metrics in the case of disk-overflow and write-through right?

Reply:
I'm not 100% sure when tracking metrics should be accepted. If telegraf goes down for whatever reason and there are some metrics written to WAL files, is it meaningful at all to accept a tracking metric in a relaunch? Or is it just an in-memory check for the current runtime? My understanding is the latter, so I think any time metrics are written to WAL or to output, tracking metrics should be accepted (and I sort of trailed off on this sentence, so I'll amend it)

Reply:
I think we should accept the metric as soon as it is written to disk because we will send the metric either some time later or after a restart of Telegraf...